certbot-apache/certbot_apache/_internal/override_centos.py

```diff
@@ -72,7 +72,20 @@ def _prepare_options(self) -> None:
         super()._prepare_options()
         if not self.options.restart_cmd_alt:  # pragma: no cover
             raise ValueError("OS option restart_cmd_alt must be set for CentOS.")
-        self.options.restart_cmd_alt[0] = self.options.ctl
+
+        # Make sure new versions (>=9) of RHEL and other derived distros (that
+        # track RHEL's versioning scheme) use the httpd executable
+        os_name, os_version = util.get_os_info()
+        rhel_derived = os_name not in [
+            "fedora",
+            "amazon", # Amazon has a yearly release version
+        ]
+        new = util.parse_loose_version(os_version) >= util.parse_loose_version('9')
+        if rhel_derived and new:
+            self.options.ctl = 'httpd'
```

Comment:
I think this sets the value of `--apache-ctl` (see `certbot --help apache`) with no way for the user to override it while continuing to list the default value as `apachectl` in the help output on RHEL 9+. Is that right?

If so,  I think we may want to do something like what we do [here](https://github.com/certbot/certbot/blob/a845ab844622a5419de166a3a35bb4dca33d8060/certbot-apache/certbot_apache/_internal/entrypoint.py#L47-L54) and define a new override class with static defaults.

What do you think?

Reply:
true, though i was hoping to leave all the finicky distro-specific details out of `get_configurator()` and instead handle it within `CentOSConfigurator`.

one issue w/ relying on static defaults is that `_prepare_options()` always sets `options.restart_cmd`'s binary to `options.ctl` (https://github.com/certbot/certbot/blob/master/certbot-apache/certbot_apache/_internal/configurator.py#L184-L187), so even if our CentOS 9 defaults have `httpd` for one and `apachectl` for the other, they'll both end up as `httpd` unless we intervene. what about something like:
```
diff --git a/certbot-apache/certbot_apache/_internal/configurator.py b/certbot-apache/certbot_apache/_internal/configurator.py
index 82b3dcaec..8d7d4eb3a 100644
--- a/certbot-apache/certbot_apache/_internal/configurator.py
+++ b/certbot-apache/certbot_apache/_internal/configurator.py
@@ -178,14 +178,14 @@ class ApacheConfigurator(common.Configurator):
             # Config options use dashes instead of underscores
             if self.conf(o.replace("_", "-")) is not None:
                 setattr(self.options, o, self.conf(o.replace("_", "-")))
+                # If the user overrides apachectl, update the various cmds as well
+                if o == "ctl":
+                    self.options.version_cmd[0] = self.options.ctl
+                    self.options.restart_cmd[0] = self.options.ctl
+                    self.options.conftest_cmd[0] = self.options.ctl
             else:
                 setattr(self.options, o, getattr(self.OS_DEFAULTS, o))
 
-        # Special cases
-        self.options.version_cmd[0] = self.options.ctl
-        self.options.restart_cmd[0] = self.options.ctl
-        self.options.conftest_cmd[0] = self.options.ctl
```

Reply:
Hmm. Digging into this further, I noticed that we also have an `--apache-bin` option:
```
$ certbot --help apache
usage: 
  certbot [SUBCOMMAND] [options] [-d DOMAIN] [-d DOMAIN] ...
...
  --apache-ctl APACHE_CTL
                        Full path to Apache control script (default: apachectl)
  --apache-bin APACHE_BIN
                        Full path to apache2/httpd binary (default: None)
```
I feel like in the long term if not now, we should make use of this here instead of telling our code that the name of `apachectl` is `httpd`. There are different ways to structure this, but I personally think we should try to keep these two options true to their intended usage and update other commands based on their values. By that I mean I'm essentially imagining logic that at least effectively does something like:
```
if self.options.version_cmd[0] == self.OS_DEFAULTS.ctl:
    self.options.version_cmd[0] = self.options.ctl
else:
    self.options.version_cmd[0] = self.options.bin
```
What do you think?

Also, on RHEL 9+ and friends, are these values from higher up in this file correct? Do we want to be using `apachectl` at all?
```
        restart_cmd=['apachectl', 'graceful'],
        restart_cmd_alt=['apachectl', 'restart'],
        conftest_cmd=['apachectl', 'configtest'],
```
I notice we leave them unmodified.

Reply:
> Also, on RHEL 9+ and friends, are these values from higher up in this file correct? Do we want to be using `apachectl` at all?

Authoritatively: yes, you should use `apachectl restart` or `graceful` to restart the daemon on all versions of RHEL, and will be supported in all future versions. This is supported as a thin wrapper around `systemctl $action httpd` to manipulate the systemd service. `apachectl configtest` should also work for all versions since it's so widely used. The  thing we don't support is passing arguments in.

Reply:
I reworked this to use two different configurator classes for the different scenarios (i.e. older CentOSy OSes vs. new RHEL-derived OSes). this ended up adding some more complexity to `get_configurator()`, but in the end i think it makes more sense this way.

Reply:
@wgreenberg, sorry for the conflicting suggestions between my first and second comments in the thread here. What do you think about what I said in that second post though at https://github.com/certbot/certbot/pull/9402#discussion_r989386296 about using the [apache_bin option](https://github.com/certbot/certbot/blob/e84271b36b530099a9c1d8de20d13148b99cb8a8/certbot-apache/certbot_apache/_internal/configurator.py#L77) instead of configuring our code to believe that the name to the Apache control script is `httpd`?